### PR TITLE
Replace the color picker component with the new version

### LINKED ISSUE #34287: Components: replace legacy `ColorPicker` with new version
Part of #34284

## Why

The new `ColorPicker` should replace the legacy version of the component

## What

This effectively means moving the code of the g2 component over to the `src` folder of the `@wordpress/components` package, completely replacing the legacy component.

We need to ensure that usages of the `ColorPicker` keep working as expected, despite the changes.

## A/C

- [ ] The new `ColorPicker` is promoted to the `src` folder, replacing the legacy component (a legacy adapter has already been created in #34014)
  - [ ] The code of the legacy `ColorPicker` is deleted
- [ ] Asses any potential breaking change introduced
- [ ] Test that the component works as expected, in continuity with the previous version:
  - [ ] In storybook
  - [ ] In the usages across the Gutenberg repository
- [ ] Introduce a changelog entry

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
Closes #34287

This PR replaces the existing color picker component with the new version.

It is a work in progress.
We are still using the deprecated props around the codebase and we should use the new ones, but for now using the deprecated props is a good way to test we are back-compatible so the change to new ones will be a separate PR.
The new version of the component relies on tiny color while the plan is to use colord so the component will be refactored in a follow-up PR.
The mobile app has its own component implementation but may be relying on web artifacts so we need to test if we are not breaking the APP.

### CODE REVIEW

File: packages/components/src/color-picker/index.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+/**
+ * Internal dependencies
+ */
+import { LegacyAdapter as ColorPicker } from './legacy-adapter';
+
+export default ColorPicker;
```
Comment (ciampo) [inline packages/components/src/color-picker/index.ts]: Let's keep the named export from the new component

```suggestion
export { LegacyAdapter as ColorPicker } from './legacy-adapter';
```

We can then modify `packages/components/src/index.js` to:

```js
export { ColorPicker } from './color-picker';
```

File: packages/components/src/color-picker/README.md [hunk 2]
```diff
@@ -1,24 +1,57 @@
 # ColorPicker
 
-Accessible color picker.
+<div class="callout callout-alert">
```
Comment (ciampo) [inline packages/components/src/color-picker/README.md]: Since the previous `ColorPicker` was not marked experimental, I wonder if we should remove this paragraph

File: packages/components/src/color-picker/styles.ts [hunk 3]
```diff
@@ -112,7 +112,7 @@ export const ColorfulWrapper = styled.div`
 `;
 
 export const DetailsControlButton = styled( Button )`
-	&&&& {
+	&&&&& {
```
Comment (ciampo) [inline packages/components/src/color-picker/styles.ts:115]: Why was this change necessary? Was this broken also before moving the component to the `src` folder?

File: packages/components/src/color-picker/stories/index.js [hunk 4]
```diff
@@ -6,25 +12,70 @@ import { useState } from '@wordpress/element';
...
  */
-import ColorPicker from '../';
+import ColorPicker from '..';
+import { Flex } from '../../flex';
+import { Spacer } from '../../spacer';
+import { space } from '../../ui/utils/space';
+
+export default {
+	component: ColorPicker,
+	title: 'Components (Experimental)/ColorPicker',
```
Comment (ciampo) [inline packages/components/src/color-picker/stories/index.js]: Same note as before about marking this component as Experimental

File: packages/components/src/color-picker/use-deprecated-props.ts [hunk 5]
```diff
@@ -69,7 +69,10 @@ function isLegacyProps( props: any ): props is LegacyProps {
 
 function getColorFromLegacyProps(
 	props: LegacyProps
-): ColorFormats.HSL | ColorFormats.HSLA {
+): ColorFormats.HSL | ColorFormats.HSLA | undefined {
+	if ( ! props.color ) {
```
Comment (ciampo) [inline packages/components/src/color-picker/use-deprecated-props.ts:73]: Looking at the `LegacyProps` type, `props.color` should be always defined.

What's the case for `color` potentially not being defined? In that case, we should update the `LegacyProps` type

**Review by ciampo — COMMENTED** [review]
Thank you for working on this, @jorgefilipecosta ! Always good to see PRs with a lot of lines removed 😄 

Since the `ColorPicker` component is now written in TypeScript, we should add its folder to `tsconfig.json`:

<details>
<summary>Click to see suggested changes</summary>

```diff
diff --git a/packages/components/tsconfig.json b/packages/components/tsconfig.json
index cd26b356e6..824f194fe2 100644
--- a/packages/components/tsconfig.json
+++ b/packages/components/tsconfig.json
@@ -27,6 +27,7 @@
 		"src/base-field/**/*",
 		"src/button/**/*",
 		"src/card/**/*",
+		"src/color-picker/**/*",
 		"src/dashicon/**/*",
 		"src/disabled/**/*",
 		"src/divider/**/*",
```

</details>

> We are still using the deprecated props around the codebase and we should use the new ones, but for now using the deprecated props is a good way to test we are back-compatible so the change to new ones will be a separate PR.

Agree 100%

> The new version of the component relies on tiny color while the plan is to use colord so the component will be refactored in a follow-up PR.

Makes sense.

> The mobile app has its own component implementation but may be relying on web artifacts so we need to test if we are not breaking the APP.

Absolutely. In general, we should do plenty of smoke testing where the `ColorPicker` component is used to make sure we're not introducing any breaking changes.

File: packages/components/src/color-picker/styles.ts (hunk 3 above)
Comment (jorgefilipecosta) [reply packages/components/src/color-picker/styles.ts:115]: The component was not broken before and worked well on storybook because the order allowed this rule to overwrite another rule with the same specificity, but when used inside the WordPress interface the code order is not the same so I needed to increase the specificity.

File: packages/components/src/color-picker/use-deprecated-props.ts (hunk 5 above)
Comment (jorgefilipecosta) [reply packages/components/src/color-picker/use-deprecated-props.ts:73]: The case of color being undefined is probably the most common one. For example, we are choosing custom a text or background color for a block by default there is no customization and that color is undefined. I added undefined to the LegacyProps props type as recommended 👍

**Comment by jorgefilipecosta** [comment]
Hi @ciampo, thank you for the review your feedback was applied.

**Review by youknowriad — COMMENTED** [review]
I kind of expected a good impact on the bundle size but the number return is weird (increased), maybe a rebase is needed to get the right result.

Are we still using deprecated props in our codebase? What's the plan for that?

Anyway, this is looking good for me.

**Review by youknowriad — APPROVED** [review]

**Review by ciampo — APPROVED** [review]
I believe we need to delete [this line](https://github.com/WordPress/gutenberg/blob/2d99d85deee70baccbe87e8b473e57bd8da78427/packages/components/src/style.scss#L9) too (and what about `style.native.scss`?)

Apart from this, changes LGTM 🚀  Thank you for working on this 

---

I see two follow-up PRs here:

- Look for usages of `ColorPicker` still using the deprecated APIs and updated them to the new ones
- refactor from `tinycolor2` to `colord`

Would you be able to take care of these too, @jorgefilipecosta ? I'd be happy to help with feedback and reviews on those ones.

**Comment by jorgefilipecosta** [comment]
Thank you for the reviews @youknowriad, @ciampo,

> I kind of expected a good impact on the bundle size but the number return is weird (increased), maybe a rebase is needed to get the right result.

I think the result is right and for now, the new component has a slightly higher bundle size. On a direct comparison, react-colorful has a smaller bundle size than react-color. I think we don't benefit from that yet because one of the reasons react-colorful has a smaller bundle size is because it does not require tinycolor, and we are still using tinycolor anyway so we don't see right away that smaller size. Also, we were using code from react-color but not including everything so our code may be smaller than what including react color would result. The new component also includes more functionality like copied timeout, hide and show details etc, and for now, we are including some styles that should be general as overwrites on the component level. I think joining these factors together explains the small increase, we should recover when tinycolor is removed.

> Are we still using deprecated props in our codebase? What's the plan for that?

Yes, we are, it was a way to test we are back-compatible. I'm working on a follow-up PR to update the codebase.

> I see two follow-up PRs here:
> Would you be able to take care of these too, @jorgefilipecosta? I'd be happy to help with feedback and reviews on those ones.

Yes, I'm working on those follow-ups, thank you for the availability to review, I will let you know when they are ready.
